roachtest: sqlsmith/setup=rand-tables/setting=no-ddl failed

roachtest.sqlsmith/setup=rand-tables/setting=no-ddl [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2924709&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2924709&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-ddl) on master @ [2dbd33ce319439bc424a52896a03e0a7ccff07f1](https://github.com/cockroachdb/cockroach/commits/2dbd33ce319439bc424a52896a03e0a7ccff07f1):


```
		WITH
			with_929 (col_6475)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(date_trunc(NULL::STRING, '2017-01-09 21:37:33.000457':::TIMESTAMP::TIMESTAMP)::TIMESTAMP),
								('2006-02-05 15:17:05.000787':::TIMESTAMP),
								('1990-06-07 17:41:00.000275':::TIMESTAMP)
						)
							AS tab_2981 (col_6475)
					INTERSECT SELECT * FROM (VALUES ('2023-11-02 16:15:59.000647':::TIMESTAMP)) AS tab_2982 (col_6476)
				)
		SELECT
			count(*) AS col_6477
		FROM
			defaultdb.public.table3@[0] AS tab_2983
		GROUP BY
			tab_2983.col3_5, tab_2983.col3_8, tab_2983.col3_9
		LIMIT
			4:::INT8;

	cluster.go:1716,context.go:140,cluster.go:1705,test_runner.go:858: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2924709-1619463598-13-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 2: 10475
		3: 10396
		4: 10025
		1: dead
		Error: UNCLASSIFIED_PROBLEM: 1: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1162
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:284
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:856
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:960
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:897
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:2057
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:204
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (3) 1: dead
		Error types: (1) errors.Unclassified (2) *withstack.withStack (3) *errutil.leafError
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh sqlsmith/setup=rand-tables/setting=no-ddl
```
</p>
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-ddl.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

This last one looks like a different issue:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xeef862]

goroutine 33922 [running]:
panic(0x4620160, 0x7e481a0)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc001284ee8 sp=0xc001284e20 pc=0x48b725
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1(0xc001285a58)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58 +0x68b fp=0xc001285088 sp=0xc001284ee8 pc=0x1d1c5cb
runtime.call32(0x0, 0x53fd3a8, 0xc003af84f8, 0x800000008)
	/usr/local/go/src/runtime/asm_amd64.s:540 +0x3e fp=0xc0012850b8 sp=0xc001285088 pc=0x4c30be
runtime.reflectcallSave(0xc0012851f8, 0x53fd3a8, 0xc003af84f8, 0x8)
	/usr/local/go/src/runtime/panic.go:881 +0x58 fp=0xc0012850e8 sp=0xc0012850b8 pc=0x48b118
runtime.runOpenDeferFrame(0xc00119f200, 0xc003af84b0, 0xc001285240)
	/usr/local/go/src/runtime/panic.go:855 +0x2cd fp=0xc001285178 sp=0xc0012850e8 pc=0x48afcd
panic(0x4620160, 0x7e481a0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9 fp=0xc001285240 sp=0xc001285178 pc=0x48b399
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:212
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:742 +0x413 fp=0xc001285270 sp=0xc001285240 pc=0x4a2473
github.com/cockroachdb/cockroach/pkg/util/tracing.SpanFromContext(0x0, 0x0, 0x7f883ddc07d0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/context.go:25 +0x22 fp=0xc0012852a8 sp=0xc001285270 pc=0xeef862
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog(0x0, 0x0, 0x3, 0xc00090d400, 0xc001285368)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:94 +0x39 fp=0xc0012852e0 sp=0xc0012852a8 pc=0x1018ff9
github.com/cockroachdb/cockroach/pkg/util/log.vEventf(0x0, 0x0, 0xc000520b00, 0x1, 0x2, 0x4e54771, 0x57, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:242 +0x12a fp=0xc001285410 sp=0xc0012852e0 pc=0x101a3ea
github.com/cockroachdb/cockroach/pkg/util/log.VEvent(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:263
github.com/cockroachdb/cockroach/pkg/sql/execinfra.MisplannedRanges(0x0, 0x0, 0xc001f47650, 0x1, 0x1, 0x4, 0xc000b6aba0, 0xc001285788, 0x4df73c, 0x7e55520)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/readerbase.go:49 +0x8c fp=0xc001285728 sp=0xc001285410 pc=0x246e0ec
github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*ColBatchScan).DrainMeta(0xc0013c4400, 0x5b7c480, 0xc0008c7ad0, 0x4a8e230d5e, 0x8893d20, 0x7a774)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/colbatch_scan.go:116 +0x677 fp=0xc001285980 sp=0xc001285728 pc=0x3388317
github.com/cockroachdb/cockroach/pkg/sql/colexecop.MetadataSources.DrainMeta.func1()
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecop/operator.go:385 +0xa8 fp=0xc001285a20 sp=0xc001285980 pc=0x29d8d68
github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0xc001285ab8, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91 +0x57 fp=0xc001285a50 sp=0xc001285a20 pc=0x1d1ba77
github.com/cockroachdb/cockroach/pkg/sql/colexecop.MetadataSources.DrainMeta(0xc001c5c6c0, 0x1, 0x6, 0x5b7c480, 0xc0008c7ad0, 0xc002387c60, 0x1, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecop/operator.go:383 +0xd2 fp=0xc001285b68 sp=0xc001285a50 pc=0x29d8ad2
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).sendMetadata(0xc001f44820, 0x5b7c480, 0xc0008c7ad0, 0x7f88158f1788, 0xc003dc63c0, 0x5ada320, 0xc003dc64a0, 0xc001285e08, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:324 +0x28f fp=0xc001285da0 sp=0xc001285b68 pc=0x2aa858f
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).runWithStream(0xc001f44820, 0x5b7c480, 0xc0008c7ad0, 0x7f88158f1788, 0xc003dc63c0, 0xc001df4560, 0xc003dc6270)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:378 +0x22d fp=0xc001285e38 sp=0xc001285da0 pc=0x2aa8d0d
github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc.(*Outbox).Run(0xc001f44820, 0x5b7c3c0, 0xc001465780, 0x5ad8b80, 0xc0001151d0, 0xf9612a6700000001, 0x2e8db9ba4341ba95, 0xf0000c04b15aecc, 0x1, 0xc001df4560, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc/outbox.go:205 +0x432 fp=0xc001285f10 sp=0xc001285e38 pc=0x2aa7e52
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRemoteOutputStream.func1(0x5b7c3c0, 0xc001465780, 0xc001df4560)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:674 +0xd5 fp=0xc001285f90 sp=0xc001285f10 pc=0x33b06b5
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1(0xc0021cb0b0, 0x5b7c3c0, 0xc001465780, 0xc001df4560, 0xc000cea8c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1233 +0x44 fp=0xc001285fb8 sp=0xc001285f90 pc=0x33b27e4
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc001285fc0 sp=0xc001285fb8 pc=0x4c4ae1
created by github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1232 +0x6f
```

--- a/pkg/sql/colencoding/key_encoding.go
+++ b/pkg/sql/colencoding/key_encoding.go
@@ -33,8 +33,7 @@ import (
 // tenant id and first table id / index id prefix removed. If matches is false,
 // the key is from a different table, and the returned remainingKey indicates a
 // "seek prefix": the next key that might be part of the table being searched
-// for. The input key will also be mutated if matches is false. See the analog
-// in sqlbase/index_encoding.go.
+// for. See the analog in sqlbase/index_encoding.go.
 func DecodeIndexKeyToCols(
 	da *rowenc.DatumAlloc,
 	vecs []coldata.Vec,
@@ -67,7 +66,8 @@ func DecodeIndexKeyToCols(
 					// We don't match. Return a key with the table ID / index ID we're
 					// searching for, so the caller knows what to seek to.
 					curPos := len(origKey) - len(key)
-					key = rowenc.EncodePartialTableIDIndexID(origKey[:curPos], ancestor.TableID, ancestor.IndexID)
+					// Prevent unwanted aliasing on the origKey by setting the capacity.
+					key = rowenc.EncodePartialTableIDIndexID(origKey[:curPos:curPos], ancestor.TableID, ancestor.IndexID)
 					return key, false, false, nil
 				}
 			}
@@ -93,7 +93,8 @@ func DecodeIndexKeyToCols(
 				// We're expecting an interleaved sentinel but didn't find one. Append
 				// one so the caller can seek to it.
 				curPos := len(origKey) - len(key)
-				key = encoding.EncodeInterleavedSentinel(origKey[:curPos])
+				// Prevent unwanted aliasing on the origKey by setting the capacity.
+				key = encoding.EncodeInterleavedSentinel(origKey[:curPos:curPos])
 				return key, false, false, nil
 			}
 		}
@@ -106,7 +107,8 @@ func DecodeIndexKeyToCols(
 			// We don't match. Return a key with the table ID / index ID we're
 			// searching for, so the caller knows what to seek to.
 			curPos := len(origKey) - len(key)
-			key = rowenc.EncodePartialTableIDIndexID(origKey[:curPos], desc.GetID(), index.GetID())
+			// Prevent unwanted aliasing on the origKey by setting the capacity.
+			key = rowenc.EncodePartialTableIDIndexID(origKey[:curPos:curPos], desc.GetID(), index.GetID())
 			return key, false, false, nil
 		}
 	}
@@ -125,7 +127,8 @@ func DecodeIndexKeyToCols(
 	// table.
 	if _, ok := encoding.DecodeIfInterleavedSentinel(key); ok {
 		curPos := len(origKey) - len(key)
-		key = encoding.EncodeNullDescending(origKey[:curPos])
+		// Prevent unwanted aliasing on the origKey by setting the capacity.
+		key = encoding.EncodeNullDescending(origKey[:curPos:curPos])
 		return key, false, false, nil
 	}
 

